internal/handlers/common/projection.go

```diff
@@ -26,78 +26,148 @@ import (
-				if !slices.Contains(supportedProjectionTypes, projectionType) {
-					err = lazyerrors.Errorf("projection of %s is not supported", projectionType)
-					return
-				}
 
-				switch projectionType {
-				case "$elemMatch":
-					inclusion = true
-				case "$slice":
-					inclusion = false
-				default:
-					panic(projectionType + " not supported")
-				}
+			inclusion, exclusion, err = validateDocProjectionExpression(v, depth, inclusion, exclusion)
+
+		case *types.Array:
+			err = NewErrorMsg(ErrElemMatchObjectRequired, "elemMatch: Invalid argument, object required, but got array")
+			return false, false, err
+
+		default: // scalar
+			if k == projectionElemMatch {
+				err = NewError(ErrElemMatchObjectRequired,
+					fmt.Errorf("elemMatch: Invalid argument, object required, but got %T", v),
+				)
+				return false, false, err
 			}
 
-		case float64, int32, int64:
-			if types.Compare(v, int32(0)) == types.Equal {
-				if inclusion {
-					err = NewError(ErrProjectionExIn,
-						fmt.Errorf("Cannot do exclusion on field %s in inclusion projection", k),
-					)
-					return
-				}
-				exclusion = true
-			} else {
-				if exclusion {
-					err = NewError(ErrProjectionInEx,
-						fmt.Errorf("Cannot do inclusion on field %s in exclusion projection", k),
-					)
-					return
-				}
-				inclusion = true
+			inclusion, exclusion, err = validateScalarProjectionExpression(v, k, inclusion, exclusion)
+		}
+	}
+	return inclusion, exclusion, err
+}
+
+func validateScalarProjectionExpression(v any, field string, inclusion, exclusion bool) (bool, bool, error) {
+	var err error
+	switch v := v.(type) {
+	case float64, int32, int64:
+		if types.Compare(v, int32(0)) == types.Equal {
+			if inclusion {
+				err = NewError(ErrElemMatchExclusionInInclusion,
+					fmt.Errorf("Cannot do exclusion on field %s in inclusion projection", field),
+				)
+				return false, false, err
 			}
+			exclusion = true
+		} else {
+			if exclusion {
+				err = NewError(ErrElemMatchInclusionInExclusion,
+					fmt.Errorf("Cannot do inclusion on field %s in exclusion projection", field),
+				)
+				return false, false, err
+			}
+			inclusion = true
+		}
 
-		case bool:
-			if v {
-				if exclusion {
-					err = NewError(ErrProjectionInEx,
-						fmt.Errorf("Cannot do inclusion on field %s in exclusion projection", k),
-					)
-					return
-				}
+	case bool:
+		if v {
+			if exclusion {
+				err = NewError(ErrElemMatchInclusionInExclusion,
+					fmt.Errorf("Cannot do inclusion on field %s in exclusion projection", field),
+				)
+				return false, false, err
+			}
+			inclusion = true
+		} else {
+			if inclusion {
+				err = NewError(ErrElemMatchExclusionInInclusion,
+					fmt.Errorf("Cannot do exclusion on field %s in inclusion projection", field),
+				)
+				return false, false, err
+			}
+			exclusion = true
+		}
+	default:
+		err = NewError(ErrNotImplemented, fmt.Errorf("%v of (%T) is not supported", v, v))
+		return inclusion, exclusion, err
```

Comment:
Should we return `false, false, err` as in other cases?

Reply:
it's just because here is an error and we have to return a default value.

Reply:
fixed.